### PR TITLE
Import

### PR DESCRIPTION
**Pull request by etokareva**

### CODE REVIEW

File: demo/pages/import/index.html [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<q-grid model="$ctrl.model"
+        edit-mode="cell"
+        selection-mode="range">
+	<q-grid:columns ng-if="$ctrl.model.data().rows.length">
```
Comment (titovmx) [inline demo/pages/import/index.html]: Why do we need in this ng-if expression?
Besides, I don't see that ```model``` put in ```$ctrl```

File: src/core/export/csv.js [hunk 2]
```diff
@@ -15,8 +15,10 @@ export class Csv {
 		const values = [];
 		let head = [];
 		for (let column of columns) {
-			values.push(valueFactory(column));
-			head.push(escape(column.title));
+			if (column.key !== 'rowNumber') {
```
Comment (titovmx) [inline src/core/export/csv.js]: Maybe check type will be better. 
And what's about row-indicator column? Should it be here too?

File: src/core/import/csv.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+function arrayToObj(array) {
```
Comment (klumba12) [inline src/core/import/csv.js]: please add tests for that file

File: src/core/import/xml.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
+const types = {
```
Comment (klumba12) [inline src/core/import/xml.js]: add tests

File: src/core/navigation/navigation.view.js [hunk 5]
```diff
@@ -15,15 +15,23 @@ export class NavigationView extends View {
 		this.shortcutOff = shortcut.register('navigation', navigation.commands);
 
 		this.blur = new Command({
-			execute: (row, column) => table.body.cell(row, column).removeClass(`${GRID_PREFIX}-focus`),
+			execute: (row, column) =>{
+				table.body.cell(row, column).removeClass(`${GRID_PREFIX}-focus`);
```
Comment (klumba12) [inline src/core/navigation/navigation.view.js]: this should be done outside the grid in excel example, if it's not possible we will add this future,
for now try to use style API

File: src/core/services/title.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+function firstRowTitle(index, row) {
+	return row[index];
+}
+function numericTitle(index) {
+	return index;
+}
+function alphabeticalTitle(index) {
```
Comment (klumba12) [inline src/core/services/title.js]: alphaTitle is ok

File: src/plugins/import/import.html [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<div class="q-grid-export">
```
Comment (klumba12) [inline src/plugins/import/import.html]: q-grid-export?

File: src/plugins/import/import.html [hunk 8]
```diff
@@ -0,0 +1,3 @@
...
+	<button type="button" ng-if="$import.upload.canExecute()">Upload file</button>
```
Comment (titovmx) [inline src/plugins/import/import.html]: Probably ```ng-click``` should be here

File: demo/pages/import/index.html (hunk 1 above)
Comment (etokareva) [reply demo/pages/import/index.html]: thanks, forgot to remove

File: src/core/export/csv.js (hunk 2 above)
Comment (etokareva) [reply src/core/export/csv.js]: Maybe.
No, it shouldn't

Comment (titovmx) [reply src/core/export/csv.js]: Another one service column is ```pad``` column. Will it be exported?

File: src/plugins/import/import.html (hunk 8 above)
Comment (etokareva) [reply src/plugins/import/import.html]: yes, this is mistake)

File: src/core/export/csv.js (hunk 2 above)
Comment (etokareva) [reply src/core/export/csv.js]: Theme with columns for export should be discussed more, I think

File: src/core/column-list/column.list.generate.js [hunk 9]
```diff
@@ -1,15 +1,16 @@
 import {startCase} from '../utility';
 import {compile, getType} from '../services';
 
-export function generate(rows, columnFactory, deep = true) {
-	if (!rows || rows.length === 0) {
+export function generate(context) {
+	const deep = context.deep || true;
+	if (!context.rows || context.rows.length === 0 || !context.columnFactory) {
```
Comment (klumba12) [inline src/core/column-list/column.list.generate.js]: please use default object for context

File: demo/pages/import/index.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+<thead>
+	<tr>
+		<th>Parameter</th>
+		<th>Type</th>
+		<th>Description</th>
+	</tr>
+</thead>
+<tbody>
+    <tr>
+        <td>header</td>
```
Comment (klumba12) [inline demo/pages/import/index.md]: head: 'alpha' | 'numeric' | 'default'

File: src/core/column-list/column.list.generate.js (hunk 9 above)
Comment (klumba12) [reply src/core/column-list/column.list.generate.js]: title by default shouldBe startCase
signature is: title(propertyPath, graph, index)

File: src/core/export/csv.js (hunk 2 above)
Comment (klumba12) [reply src/core/export/csv.js]: column has a class property, you can export only class === 'data' for example

File: src/core/import/csv.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	const result = {};
+	for (let i = 0; i < array.length; i++) {
+		result['' + i] = array[i];
+	}
+	return result;
+}
+
+function traverse(row) {
+	const result = [];
```
Comment (klumba12) [inline src/core/import/csv.js]: need to refactor this

File: src/core/import/json.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import {isArray} from '@grid/core/utility';
+
+export class Json {
+	read(data) {
+		const rows = JSON.parse(data);
+		if (isArray(rows)) {
+			return rows;
+		}
+		return [];
```
Comment (klumba12) [inline src/core/import/json.js]: [rows]

File: src/core/import/xml.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	return children.length > 1;
+}
+function isElement(xml) {
+	return xml.nodeType === types.ELEMENT_NODE;
+}
+function isTextContainer(xml) {
+	return xml.nodeType === types.ELEMENT_NODE && !xml.children.length && xml.childNodes.length;
+}
+
+function generateArray(child) {
```
Comment (klumba12) [inline src/core/import/xml.js]: not readable name, cant understand what it does

File: src/core/import/xml.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+			result.push(obj);
+		}
+		child.setAttribute('visited', '');
+	}
+	return result;
+}
+function isVisited(xml) {
+	return xml.getAttribute('visited');
+}
+function parse(xml, obj, tagName) {
```
Comment (klumba12) [inline src/core/import/xml.js]: the same parse doesn't return anything so don't understand what it does

File: src/plugins/import/import.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+				const input = doc.createElement('input');
+				input.type = 'file';
+				input.className = 'ng-hide';
+				element.appendChild(input);
+				input.click();
+			}
+		});
+	}
+
+	handleFile(e) {
```
Comment (klumba12) [inline src/plugins/import/import.js]: rename it

File: src/plugins/export/pdf.js [hunk 16]
```diff
@@ -1,6 +1,6 @@
 import jsPDF from 'jspdf';
```
Comment (klumba12) [inline src/plugins/export/pdf.js]: don't import this files directly, use global namespace for them, also add them to vender list in webpack config.
now they are written to dist, also in documention please add that user should import this vendor libraries

File: src/core/export/csv.js (hunk 2 above)
Comment (titovmx) [reply src/core/export/csv.js]: ```class``` property is really good. But maybe it should be defined as 'data' in DataColumnModel? I'm confused a bit that it's declared in base class

File: src/core/import/xml.js (hunk 13 above)
Comment (etokareva) [reply src/core/import/xml.js]: it creates array

File: src/core/import/xml.js (hunk 14 above)
Comment (etokareva) [reply src/core/import/xml.js]: it changes the result object, which we pass from the outside using xml object

File: src/plugins/export/pdf.js (hunk 16 above)
Comment (etokareva) [reply src/plugins/export/pdf.js]: it's already in webpack config. It has been added in my previous pull request

File: src/plugins/import/import.js (hunk 15 above)
Comment (etokareva) [reply src/plugins/import/import.js]: 'load ' will be okay?

File: src/core/import/json.js (hunk 12 above)
Comment (etokareva) [reply src/core/import/json.js]: this won't work in case of return [rows] if rows is array

File: src/core/column-list/column.list.generate.js (hunk 9 above)
Comment (etokareva) [reply src/core/column-list/column.list.generate.js]: which object is default?

File: src/core/export/csv.js (hunk 2 above)
Comment (etokareva) [reply src/core/export/csv.js]: Actually, I found out that for column type: "row-number" class is equal to "data"

File: demo/pages/excel/index.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+				rows.push(row);
+			}
+
+			next(rows);
+		}]
+			.concat(qgrid.pipeUnit.view)
+	});
+
+	model.style({
+		// row: (row, context) => {
```
Comment (klumba12) [inline demo/pages/excel/index.js]: please, remove these comments

File: src/core/column-list/column.list.generate.js [hunk 18]
```diff
@@ -1,15 +1,21 @@
...
-	if (!rows || rows.length === 0) {
-		return [];
+export function generate(settings) {
+	const context = assignWith({
+		deep: true,
+		rows: [],
+		columnFactory: () => new TextColumnModel(),
+		title: startCase
+	}, settings);
+	if (context.rows && context.rows.length) {
```
Comment (klumba12) [inline src/core/column-list/column.list.generate.js]: don't need to check context.rows

File: demo/pages/excel/index.js (hunk 17 above)
Comment (vkorolev) [inline demo/pages/excel/index.js]: Consider to uncomment or delete this snippet

File: src/core/export/export.service.js [hunk 19]
```diff
@@ -23,5 +23,5 @@ function flattenObject(obj, separator = ', ') {
 }
 
 export {
-	flattenObject
+	flatView
```
Comment (vkorolev) [inline src/core/export/export.service.js:26]: **export** service exposes single function called `flatView`. Consider to rename file to flatten or like this

File: src/core/import/csv/csv.spec.js [hunk 20]
```diff
@@ -0,0 +1,34 @@
+import {Csv} from './csv';
+
+const csv = new Csv();
+
+describe('csv parser', () => {
+	it('should return empty array', () => {
```
Comment (vkorolev) [inline src/core/import/csv/csv.spec.js]: should return empty string when input is not defined

Comment (etokareva) [reply src/core/import/csv/csv.spec.js]: I think it should return array of rows

File: src/core/import/csv/csv.spec.js [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+		expect(csv.read()).to.eql([]);
+	});
+	it('should return empty array when passed empty string', () => {
+		expect(csv.read('')).to.eql([]);
+	});
+
+	it('should return array of rows', () => {
+		const text = `Name First,Name Last,Gender,Birthday,Contact Address Street,Contact Address Zip,Contact Address City,Contact Address State,Contact Email,Contact Region,Contact Phone,Likes,Member Since
```
Comment (vkorolev) [inline src/core/import/csv/csv.spec.js]: imo this test data is too verbose. i think 2x2 table would be enough

Comment (etokareva) [reply src/core/import/csv/csv.spec.js]: ok

File: src/core/export/json.js [hunk 22]
```diff
@@ -14,34 +14,17 @@ function rewriteHeaders(rows, titles) {
...
-
-			return rewriteHeaders(result, titles);
+		for (let row of rows) {
+			result.push(flatView(row));
+		}
+		for (let column of columns) {
+			titles.push(column.title);
 		}
 
+		return rewriteHeaders(result, titles);
```
Comment (klumba12) [inline src/core/export/json.js]: rename it to updateTitles

File: src/core/import/csv/csv.spec.js (hunk 20 above)
Comment (vkorolev) [reply src/core/import/csv/csv.spec.js]: sorry, of course.
i meant that you should have more detailed sentence 
`'should return empty array wnhen input is not defined'`

File: src/core/import/csv/csv.spec.js (hunk 21 above)
Comment (vkorolev) [reply src/core/import/csv/csv.spec.js]: It's better to have multiple small tests which would check parser's edge cases

File: src/core/export/xml.js [hunk 23]
```diff
@@ -53,22 +34,14 @@ function objToXml(obj) {
...
-	headers.push('</head>');
-	return headers.join('');
-}
 
 export class Xml {
-	write(rows, columns) {
-		const result = [`${begin}${schema}`, ...formHead(columns), '<body>'];
+	write(rows) {
+		const result = [begin];
 		for (let row of rows) {
```
Comment (klumba12) [inline src/core/export/xml.js:41]: rows.map(objToXml) looks pretty doesn't it? :)

File: src/plugins/import/import.js [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+const Plugin = PluginComponent('import');
+class Import extends Plugin {
+	constructor() {
+		super(...arguments);
+		const element = this.$element[0];
+		const doc = element.ownerDocument;
+		this.eventListener = new EventListener(element, new EventManager(this));
+		this.upload = new Command({
+			execute: () => {
+				const input = doc.createElement('input');
```
Comment (klumba12) [inline src/plugins/import/import.js]: can we have in core upload.js with that code?

File: src/core/import/csv/csv.js [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+
+				continue;
+			}
+
+			if (c === CharReader.eof) {
+				reader.read();
+
+				if (line.length > 0 || term.length > 0) {
+					line.push(term);
+					if (line.length > 0) {
```
Comment (titovmx) [inline src/core/import/csv/csv.js]: I think this condition is redundant

File: src/core/import/xml/xml.spec.js [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+		expect(xml.read()).to.eql([]);
+	});
+
+	it('should return empty array when passed empty string', () => {
+		expect(xml.read('')).to.eql([]);
+	});
+
+	it('should return array of rows', () => {
+		const text = `<?xml version="1.0" encoding="UTF-8"?>
+		<root>
```
Comment (titovmx) [inline src/core/import/xml/xml.spec.js:30]: can it be placed to separate file or reduced?

File: src/view/components/head/head.html [hunk 27]
```diff
@@ -1,18 +1,19 @@
...
+	    q-grid:on-drop="$view.head.drop.execute($event)"
+	    q-grid:can-drop="$view.head.drop.canExecute($event)"
+	    q-grid:can-drag="$view.head.drag.canExecute($event)"
+	    q-grid:resizable="$view.head.resize.canExecute($event)"
+	    q-grid:resize="$column.model.key"
+	    q-grid:resize-path="columns"
+	    q-grid:can-resize="$view.head.resize.canExecute($event)"
+	    q-grid-core:th>
 	</th>
+	<th ng-if="$view.head.rows.length === 1 && $view.body.rows.length === 0"></th>
```
Comment (titovmx) [inline src/view/components/head/head.html]: why do we need in this ```th```?

File: src/core/export/export.service.js (hunk 19 above)
Comment (klumba12) [reply src/core/export/export.service.js:26]: Its ok name i think no need to make separate file in this case

File: src/view/components/head/head.html (hunk 27 above)
Comment (klumba12) [reply src/view/components/head/head.html]: +1

Comment (etokareva) [reply src/view/components/head/head.html]: Actually, now we don't need this :-)

File: src/core/export/xml.js (hunk 23 above)
Comment (etokareva) [reply src/core/export/xml.js:41]: I don't know how to wrap object in row tag in that case

File: src/core/import/csv/csv.js (hunk 25 above)
Comment (etokareva) [reply src/core/import/csv/csv.js]: I think you are right

File: src/core/import/xml/xml.spec.js (hunk 26 above)
Comment (etokareva) [reply src/core/import/xml/xml.spec.js:30]: reduced

**Review by klumba12 — APPROVED** [review]
